engine/src/main/java/com/datatorrent/stram/util/JarHelper.java

```diff
@@ -0,0 +1,168 @@
+          } catch (IOException e) {
+            throw new AssertionError("Cannot resolve jar file for " + jarClass, e);
+          }
+        }
+        if ("file".equals(location.getProtocol())) {
+          jar = location.getFile();
+          final File dir = new File(jar);
+          if (dir.isDirectory()) {
+            try {
+              File temp = File.createTempFile("apex-", ".jar");
+              new JarHelper(temp).createJar(dir);
+              jar = temp.getAbsolutePath();
+            } catch (IOException e) {
+              throw new AssertionError("Cannot resolve jar file for " + jarClass + ". URL " + location, e);
+            }
+          }
+        } else {
+          throw new AssertionError("Cannot resolve jar file for " + jarClass + ". URL " + location);
+        }
+        sourceToJar.put(location, jar);
+        logger.debug("added sourceLocation {} as {}", location, jar);
+      }
+      if (jar == null) {
+        throw new AssertionError("Cannot resolve jar file for " + jarClass);
+      }
+    }
+    return jar;
+  }
+
+  private final ArchiveOutputStream aos;
+
+  private JarHelper(File file) throws IOException
+  {
+    try {
+      aos = new ArchiveStreamFactory().createArchiveOutputStream(ArchiveStreamFactory.JAR, new FileOutputStream(file));
+    } catch (ArchiveException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private void createJar(File dir) throws IOException
+  {
+    File manifestFile = new File(dir, JarFile.MANIFEST_NAME);
+    if (!manifestFile.exists()) {
+      aos.putArchiveEntry(new JarArchiveEntry(JarFile.MANIFEST_NAME));
+      new Manifest().write(aos);
+      aos.closeArchiveEntry();
+    } else {
+      addEntry(manifestFile, JarFile.MANIFEST_NAME);
+    }
+    final java.nio.file.Path root = dir.toPath();
+    Files.walkFileTree(root,
+        new SimpleFileVisitor<Path>()
+        {
+          String relativePath;
+
+          @Override
+          public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException
+          {
+            relativePath = root.relativize(dir).toString();
+            if (!relativePath.isEmpty()) {
+              if (!relativePath.endsWith("/")) {
+                relativePath += "/";
+              }
+              addEntry(dir.toFile(), relativePath);
+            }
+            return super.preVisitDirectory(dir, attrs);
+          }
+
+          @Override
+          public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException
+          {
+            String name = relativePath + file.getFileName().toString();
+            if (!JarFile.MANIFEST_NAME.equals(name)) {
+              addEntry(file.toFile(), relativePath + file.getFileName().toString());
+            }
+            return super.visitFile(file, attrs);
+          }
+
+          @Override
+          public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException
+          {
```

Comment:
It is required to adjust back relativePath, otherwise, it will continue to point to visited sub-directory.